### PR TITLE
Populate dashpages from configuration

### LINKED ISSUE #5289: Populate dashpages from configuration
Initially, the set of dashpages will come from the dashboard's configuration. We should populate the relevant view models from that configuration when needed.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #5289

This change causes the dashboard to look for a file named `dashpages.json` in the current directory. If found, it's parsed to provide the set of dashpages shown on the metrics page.

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [x] No. Follow-up changes expected. (this is a PR to a feature branch)
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No (not yet, we will doc dashpages entirely before the feature branch merges)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5302)

### CODE REVIEW

File: tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                            "instrument": "envoy.connection.count",
+                            "required": true,   // non-default
+                            "resource": "envoy" // non-default
+                        }
+                    ]
+                }
+            ]
+            """;
+
+        var dashpages = DashpageJsonFilePersistence.Deserialize(json);
```
Comment (radical) [inline tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs:57]: Can the test use ` DashpageJsonFilePersistence.GetDashpagesAsync` instead since that is the public API?

Comment (drewnoakes) [reply tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs:57]: That requires writing a file or mocking the file provider which I wanted to avoid.

Comment (radical) [reply tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs:57]: Writing a json file to be loaded by the public api? It can have the same contents as in this existing test, right?

File: src/Aspire.Dashboard/Model/DashpageModels.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+            if (fileInfo.Exists == false)
+            {
+                _dashpages = [];
+            }
+            else
+            {
+                using Stream stream = fileInfo.CreateReadStream();
+
+                using StreamReader reader = new(stream, Encoding.UTF8, leaveOpen: true);
```
Comment (radical) [inline src/Aspire.Dashboard/Model/DashpageModels.cs:92]: Just curious - why is `leaveOpen: true` needed here?

Comment (drewnoakes) [reply src/Aspire.Dashboard/Model/DashpageModels.cs:92]: Either way would be fine. I prefer to have the `using Stream` close the stream, and the reader not actually touch the stream when disposed, but closing the stream twice would also be safe.

File: tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs (hunk 1 above)
Comment (drewnoakes) [reply tests/Aspire.Dashboard.Tests/Model/DashpageJsonFilePersistenceTests.cs:57]: The main thing I want to test is the transformation between string and object model. The other API tests more layers than needed here.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 3]
```diff
@@ -219,6 +220,9 @@ public DashboardWebApplication(
 
         builder.Services.AddScoped<DimensionManager>();
 
+        builder.Services.AddSingleton<IFileProvider>(new PhysicalFileProvider(Directory.GetCurrentDirectory()));
```
Comment (adamint) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:223]: why is this necessary?

File: src/Aspire.Dashboard/dashpages.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
+[
+  /* Comments are ignored */
```
Comment (adamint) [inline src/Aspire.Dashboard/dashpages.json:2]: What do you think about documenting this schema elsewhere, and removing the comments? It is a good practice to avoid commenting JSON, even though our deserializer currently handles it.

**Review by adamint — APPROVED** [review]

File: src/Aspire.Dashboard/DashboardWebApplication.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:223]: We're the first consumer of the `IFileProvider` abstraction. It allows the other service `DashpageJsonFilePersistence` access to the file system.

File: src/Aspire.Dashboard/dashpages.json (hunk 4 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/dashpages.json:2]: Comments were just because I copied this from a unit test where I wanted to validate that comments are tolerated.

Adding a schema makes sense.

**Comment by drewnoakes** [comment]
`Aspire.EndToEnd.Tests.IntegrationServicesTests.Aspire.EndToEnd.Tests.IntegrationServicesTests.VerifyCosmosComponentWorks(resourceName: efcosmos)` timed out after 10 minutes.

File: src/Aspire.Dashboard/dashpages.json (hunk 4 above)
Comment (adamint) [reply src/Aspire.Dashboard/dashpages.json:2]: I do not think we should officially support JSON with comments. It is not valid JSON.

Comment (drewnoakes) [reply src/Aspire.Dashboard/dashpages.json:2]: It's very common to want to document your configuration. It's not very friendly to disallow that when it's so easy to permit it. I see no downside.

Comment (drewnoakes) [reply src/Aspire.Dashboard/dashpages.json:2]: For example, the .NET Project System ignores comments in `launchSettings.json` files.
